akka-javasdk/src/test/java/akka/javasdk/agent/MultiAgent.java

```diff
@@ -0,0 +1,120 @@
+      
+        You can find the list of exiting agents below (in JSON format):
+        %s
+      
+        Note that each agent has a description of its capabilities. Given the user request,
+        you must define the right ordering.
+      
+        Moreover, you must generate a concise request to be sent to each agent. This agent request is of course
+        based on the user original request, but is tailored to the specific agent. Each individual agent should not receive
+        requests or any text that is not related with its domain of expertise.
+      
+        Your response should follow a strict json schema as defined bellow.
+         {
+           "steps": [
+              {
+                "agentId": "<the id of the agent>",
+                "query: "<agent tailored query>",
+              }
+           ]
+         }
+      
+        The '<the id of the agent>' should be filled with the agent id.
+        The '<agent tailored query>' should contain the agent tailored message.
+        The order og the items inside the "steps" array should be the order of execution.
+      
+        Do not include any explanations or text outside of the JSON structure.
+        """.stripIndent();
+
+    Planner(AgentRegistry registry) {
+      this.registry = registry;
+    }
+
+    public Effect<Plan> plan(String arg) {
+      var agentIds = registry.agentIdsWithRole("workers");
+      var agentDescriptions = agentIds.stream()
+          .map(registry::agentDescriptionAsJson)
+          .collect(Collectors.joining(", ", "[", "]"));
+
+      return effects()
+          .systemMessage(systemMessageTemplate.formatted(agentDescriptions))
+          .userMessage(arg)
+          .thenReplyAs(Plan.class);
+    }
+  }
+
+  @ComponentId("agent1")
+  @AgentDescription(name = "Agent 1", description = "...", role = "worker")
+  static class Agent1 extends Agent {
+    public Effect<String> call1(String arg) {
+      return effects().reply("rsp1");
+    }
+  }
+
+  @ComponentId("agent2")
+  @AgentDescription(name = "Agent 2", description = "...", role = "worker")
+  static class Agent2 extends Agent {
+    public Effect<String> call2(String arg) {
+      return effects().reply("rsp2");
+    }
+  }
+
+  @ComponentId("agent-team-a")
+  static class TeamA { // this would be a workflow
+    private final ComponentClient componentClient;
+    private final AgentRegistry registry;
+
+    TeamA(ComponentClient componentClient, AgentRegistry registry) {
+      this.componentClient = componentClient;
+      this.registry = registry;
+    }
+
+    public void run() {
+      var sessionId = UUID.randomUUID().toString();
+
+      var plan =
+        componentClient.forAgent()
+          .inSession(sessionId)
+          .method(Planner::plan)
+          .invoke("task...");
+
+      List<Object> results =
+        plan.steps.stream().map(step -> {
+          return registry.getAgent(step.agentId)
+              .invoke(step.query);
+        }).toList();
```

Comment:
This is the most difficult part. We have a plan of agent ids, and want to call them without manually mapping agent ids to Agent method lambda. Here `getAgent` returns an AgentRef, and that is essentially a ComponentMethodRef1. We could skip AgentRef and just use ComponentMethodRef1.

Anyway, the challenging part is if we want to make parameter and result typed.

Reply:
Yep, that's why in the multi-agent sample the Agent has a common interface so we know which method to call and we know it's return type.

It's not clear to me how we will be able to build a method invoker for the user? 
Also not how to preserve the types.

Reply:
I added the dynamicCall to the component client instead. Not sure about the name, but it conveys in some way that it is not typed?